### ISSUE TITLE
Adding the new parameter of the compactor for the ResLevel[12]DownsampleRange or the function of the query minSourceResolutionMillis

### ISSUE THREAD

**Issue by wotkddl21** [thanos-io/thanos#6929]
### Is your proposal related to a problem?

(Write your answer here.)
I'm using thanos compactor, query, store with the grafana.  
The compactor  
``` yaml  
        - compact
        - --log.level=debug
        - --delete-delay=24h
        - --wait
        - --wait-interval=2h
        - --objstore.config-file=/etc/objstore-config/bucketconfig.yaml
        - --retention.resolution-1h=0
        - --retention.resolution-5m=180d
        - --retention.resolution-raw=30d
```  
The query  
``` yaml  
        - query
        - --log.level=debug
        - --query.auto-downsampling
        - --query.instant.default.max_source_resolution=1h
        - --query.replica-label=prometheus_replica
        - --query.replica-label=thanos_ruler_replica
        - --store=dnssrv+_grpc._tcp.thanos-sidecar.sre-mtr.svc.cluster.local
        - --store=dnssrv+_grpc._tcp.thanos-store.sre-mtr.svc.cluster.local
        - --store=dnssrv+_grpc._tcp.thanos-ruler.sre-mtr.svc.cluster.local
```  
The store  
``` yaml   
        - store
        - --chunk-pool-size=64GB
        - --log.level=debug
        - --objstore.config-file=/etc/objstore-config/bucketconfig.yaml
        - --ignore-deletion-marks-delay=1d
        - --grpc-address=0.0.0.0:10901
        - --http-address=0.0.0.0:10902
        - --data-dir=/var/thanos/store
```  
According to above, my raw, 5m and 1h data would be stored 30days, 180days, forever respectively.  
Since I want to show the metric before 30days, I set the custom query parameter in the grafana datasource of the thanos query as max_source_resolution=1h  
And in some grafana panel, I used the rate function with vector range 2m about the network.  (2m-rate-function below)
If I change the time range in the dashboard from 30m to 7days, the panel with the 2m-rate-function shows the 40hours data only. (exactly the data before 40h are seen a little bit sad...)  
![image](https://github.com/thanos-io/thanos/assets/57653556/4e30cd6b-dc04-46eb-859b-1a0166527b3f)  
If I change the vector range from 2m to 10m (to above 5m and below 1h) it works only for the data between 40h and now.  
![image](https://github.com/thanos-io/thanos/assets/57653556/d15d006f-190c-469c-9cb9-41fd119b208f)
If I change the vector range from 10m to 2h (to above 1h) it works for it works for every data. But the graph is too rough.  

I want to get the data as detail as possible within any time range.  
### Describe the solution you'd like

(Describe your proposed solution here.)
So I think there could be 2 options.  
1. add the new cli parameter of the compactor  
2. add the new function of the query  
### Describe alternatives you've considered
1. add the new cli parameter of the compactor  
As I researched, at the compactor there are two const ResLevel1DownsampleRange, ResLevel2DownsampleRange.  
Both above are hard-coded as  40hours and 10days respectively at the thanos/pkg/compact/downsample/downsample.go#45  
How about get both time range from the user?  
If then, I could set the compactor make 5m, 1h data for the data after 30d, 180d respectively.  
Then the query would find raw, 5m data only within 30d, 180d respectively.  

2. add the new function of the query    
As I inspected, there was maxSourceResolutionMillis function at the thanos/cmd/thanos/query.go  
I think this runction return the max resolution (0, 300000, 3600000 for raw, 5m, 1h)  
How about the minSourceResolutionMills? It returns exactly the reverse value of the maxSourceResolutionMillis.  
Let assume below    (A,B,C are specific period and A is the most recent)  
raw:   A
5m:  A,B
1h:  A,B,C  
If I set min_source_resolution as 1h and query for the A,B and C timeline, then the query would return raw, 5m and 1h respectively.  
For the A period, there are raw,5m and 1h data, so query returns raw data.  
For the B period, there are 5m and 1h data, so query returns 5m data.  
For the A period, there are 1h data, so query returns 1h data.  

(Write your answer here.)

### Additional context

(Write your answer here.)

**Comment by yeya24** [comment]
Have you tried setting `$__rate_interval` in Grafana? https://grafana.com/docs/grafana/latest/datasources/prometheus/template-variables/#:~:text=We%20recommend%20using%20%24__rate_interval,avoid%20problems%20specific%20to%20Prometheus.

This could adjust the interval dynamically based on the your time range. It might not work perfectly for you but should be better than now.

**Comment by yeya24** [comment]
@wotkddl21 Do you use query frontend? Query frontend has a feature to request downsampled data. You can try it by enabling `--query-range.request-downsampled`.

**Comment by wotkddl21** [comment]
@yeya24 
I tried query frontend with --query-range.request-downsamples.
But it is same as without the frontend.  
![image](https://github.com/thanos-io/thanos/assets/57653556/acbadc7a-6718-47c0-9e72-62337e07787c)  
``` yaml  
        app: thanos-query-frontend
    spec:
      containers:
      - args:
        - query-frontend
        - --log.level=debug
        - --query-frontend.downstream-url=http://thanos-query.sre-mtr.svc.cluster.local:10902
        - --request.logging-config-file=/etc/log/config
        - --query-range.request-downsampled

```

**Comment by yeya24** [comment]
My bad, I think the query frontend doesn't work well for your case because of the range is not big enough. To get it working you need to zoom in the graph.

I don't know if there is an easy way to get it working for your setup without any code changes. There is a workaround to add template variable for the range like `rate(xxx[$range])` but it is still kind manual.  https://community.grafana.com/t/tweaking-the-interval-variable/84218/5

I don't think we will expose the downsampling configuration of compactor. It doesn't solve the root cause of the problem. Your grafana dashboard can break at some point (or break for other users) even if you change the downsampling time range.

And I don't think we can add a new query function as query engine shouldn't be aware of downsampling.

**Comment by wotkddl21** [comment]
@yeya24 OK
Then I have to split the dashboard for long and short term.
I will close this issue. Thanks

**Comment by wotkddl21** [comment]
@yeya24 You mean the custom datasource for the grafana?

**Comment by yeya24** [comment]
@wotkddl21, no. I am thinking that the custom `max_source_resolution=1h` is not flexible at the datasource level because it applies on every query.
We should drop that. Instead we have a feature in Thanos QFE to inspect your query and set `max_source_resolution` based on the input query string.

**Comment by yeya24** [comment]
https://github.com/thanos-io/thanos/pull/7012 is merged
